### PR TITLE
helper functions for adding new charts

### PR DESCRIPTION
**Pull request by jwildfire**
To test, build the package with this branch and then use the  `customSettings.R` below and start the app with `safetyGraphicsApp(settingsLocation='/path/to/customSettings/')` 

```
#customSettings.R 
# This file replaces interactive hep-explorer with a static ggplot version of the chart, and removes the group settings options. 

source('/Users/jeremy/Sites/github/safetyGraphics/inst/examples/custom_charts/custom_ggplot.R')
custom_location<-"/Users/jeremy/Sites/github/safetyGraphics/customSettings"

addSetting(
  text_key="pointcolor", 
  label="Point Color", 
  description="The color of the points", 
  setting_type="character", 
  setting_cat="appearance", 
  default="red", 
  charts="custom_ggplot", 
  settingsLocation=custom_location
)

addChart(
  chart="edishstatic",
  main="custom_ggplot", 
  label="Hep Exploprer - static",
  settingsLocation = custom_location,
  requiredSettings=c("pointcolor", "id_col", "value_col", "measure_col", "measure_values--ALT", "measure_values--TB", "group_cols", "normal_col_high", "studyday_col"),
  type="static"
)

removeCharts(
  charts="hepexplorer", 
  settingsLocation=custom_location
)

removeSettings(
  text_keys=c("group_cols"),
  settingsLocation=custom_location
)
```

### CODE REVIEW

File: R/addChart.R [hunk 1]
```diff
@@ -0,0 +1,94 @@
+#' Adds a new chart for use in the safetyGraphics shiny app
+#'
+#' This function updates settings objects to add a new chart to the safetyGraphics shiny app
+#'
+#' This function makes it easy for users to adds a new chart to the safetyGraphics shiny app, by making updates to the underlying metadata used by the package. Specifically, the function adds a row to chartsMetadata.rda describing the chart and adds a column to settingsMetadata.rda specifying which settings are used with the chart. If new settings are needed for the chart, the user should call addSetting() for each new setting required.
```
Comment (pburnsdata) [inline R/addChart.R]: change adds to add

File: R/addChart.R [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+#'
+#' @param settings_location path where the custom settings will be loaded/saved. If metadata is not found in that location, it will be read from the package (e.g. safetyGraphics::settingsMetadata), and then written to the specified location once the new chart has been added.
+#' @param chart Name of the chart - one word, all lower case
+#' @param label Nicely formatted name of the chart
+#' @param description Description of the chart
+#' @param repo_url Homepage for chart's code repository (if any)
+#' @param settings_url Homepage for chart's settings documentation
+#' @param main Name of the main function used to initialize the app. If the type is htmlwidgets, the js function must accept "location" and "settings" parameters (in that order) and have an .init() method, expecting a json data array. Otherwise, the r function should accept named data and settings parameters, andshould be loaded in the user's namespace.
```
Comment (pburnsdata) [inline R/addChart.R]: add space between and and should

File: R/addChart.R [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  # check inputs
+  stopifnot(
+    typeof(chart)=="character",
+    typeof(label)=="character",
+    typeof(description)=="character",
+    typeof(repo_url)=="character",
+    typeof(settings_url)=="character",
+    typeof(main)=="character",
+    typeof(type)=="character",
+    type %in% c("htmlwidget","plotly","static","module"),
```
Comment (pburnsdata) [inline R/addChart.R]: do we want to support modules here?

File: R/addChart.R [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+#'
+#' @param settings_location path where the custom settings will be loaded/saved. If metadata is not found in that location, it will be read from the package (e.g. safetyGraphics::settingsMetadata), and then written to the specified location once the new chart has been added.
+#' @param chart Name of the chart - one word, all lower case
+#' @param label Nicely formatted name of the chart
+#' @param description Description of the chart
+#' @param repo_url Homepage for chart's code repository (if any)
+#' @param settings_url Homepage for chart's settings documentation
+#' @param main Name of the main function used to initialize the app. If the type is htmlwidgets, the js function must accept "location" and "settings" parameters (in that order) and have an .init() method, expecting a json data array. Otherwise, the r function should accept named data and settings parameters, andshould be loaded in the user's namespace.
+#' @param type type of chart (e.g. 'htmlwidget')
```
Comment (pburnsdata) [inline R/addChart.R]: should list out potential options here in the docs

File: R/safetyGraphicsApp.R [hunk 5]
```diff
@@ -1,33 +1,70 @@
 #' Run the interactive safety graphics builder
 #'
-#' @param charts Character vector of charts to include 
+#' @param charts Character vector of charts to include
 #' @param maxFileSize maximum file size in MB allowed for file upload
-#' @param settingsLocation folder location of user-defined settings metadata
+#' @param settingsLocation folder location of user-defined settings metadata. Files should be named settingsMetadata.rda, chartsMetadata.rda and standardsMetadata.rda and use the same structure established in the /data folder.
```
Comment (bzkrouse) [inline R/safetyGraphicsApp.R]: are we going with rda or rds? Code uses Rds files but R packages store data as Rda.

File: R/safetyGraphicsApp.R [hunk 6]
```diff
@@ -1,33 +1,70 @@
...
+  }
+
+  chartsMetaPath <- paste(settingsLocation,"chartsMetadata.Rds",sep="/")
+  if(file.exists(chartsMetaPath)){
+    options(sg_chartsMetadata=TRUE) 
+    options(sg_chartsMetadata_df=readRDS(chartsMetaPath))
+    
+  } else {
+    options(sg_chartsMetadata=FALSE) 
+    options(sg_chartsMetadata_df=readRDS(chartsMetaPath))
```
Comment (bzkrouse) [inline R/safetyGraphicsApp.R]: I think this should be `options(sg_chartsMetadata_df=NULL)`?

File: R/addSetting.R [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+#' @param description Description
+#' @param setting_type Expected type for setting value. Should be "character", "vector", "numeric" or "logical"
+#' @param setting_required Flag indicating if the setting is required
+#' @param column_mapping Flag indicating if the setting corresponds to a column in the associated data
+#' @param column_type Expected type for the data column values. Should be "character","logical" or "numeric"
+#' @param field_mapping Flag indicating whether the setting corresponds to a field-level mapping in the data
+#' @param field_column_key Key for the column that provides options for the field-level mapping in the data
+#' @param setting_cat Setting category (data, measure, appearance)
+#' @param default Default value for non-data settings
+#' @param charts list of charts using this setting
```
Comment (bzkrouse) [inline R/addSetting.R]: List or character vector?

File: R/addChart.R (hunk 3 above)
Comment (jwildfire) [reply R/addChart.R]: Yeah, might not be quite as simple as the static situations. Filed an issue to cover this - #363

File: R/safetyGraphicsApp.R (hunk 5 above)
Comment (jwildfire) [reply R/safetyGraphicsApp.R]: No strong opinion really. I guess the strength of RDS is that the files are smaller ... I'm fine with changing it if anyone feels strongly. Go ahead and file an issue if you think we should change to rda.

File: R/addChart.R (hunk 3 above)
Comment (jwildfire) [reply R/addChart.R]: Per Becca in chat, seems like we're probably ok here. Will close that new issue once we document the process in the vignette.

**Pull request merged**